toolchain/base/value_store_chunk.h

```diff
@@ -0,0 +1,177 @@
+static constexpr auto PlatformChunkCapacitySize() -> int {
+  constexpr auto MaxElements =
+      PlatformChunkMaxAllocationBytes<IdT>() / sizeof(typename IdT::ValueType);
+  constexpr auto Pow2MaxElements = std::bit_ceil(MaxElements);
+  if (Pow2MaxElements > MaxElements) {
+    return Pow2MaxElements / 2;
+  }
+  return Pow2MaxElements;
+}
+
+// The number of bits needed to index each element in a chunk allocation.
+template <class IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto PlatformChunkCapacityBits() -> int {
+  static_assert(PlatformChunkCapacitySize<IdT>() > 1);
+  int bits = 0;
+  for (auto size = PlatformChunkCapacitySize<IdT>(); size > 1; size /= 2) {
+    ++bits;
+  }
+  return bits;
+}
+
+// Converts an id into an index into the set of chunks, and an offset into that
+// specific chunk.
+template <typename IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto IdToChunkIndices(IdT id) -> std::pair<int, int> {
+  constexpr auto LowBits = PlatformChunkCapacityBits<IdT>();
+
+  // Verify there are no unused bits when indexing up to the
+  // PlatformChunkCapacitySize(). This ensures that ids are contiguous values
+  // from 0, as if the values were all stored in a single array, and allows
+  // using the ids to index into other arrays.
+  static_assert((1 << LowBits) == PlatformChunkCapacitySize<IdT>());
+  // Simple check to make sure nothing went wildly wrong with the
+  // PlatformChunkCapacitySize, and we have some room for a chunk index, and
+  // that shifting by the number of bits won't be UB in an int32_t.
+  static_assert(LowBits < 24);
+
+  return {
+      id.index >> LowBits,
+      id.index & ((1 << LowBits) - 1),
+  };
+}
+
+// Converts an index into the set of chunks, and an offset into that specific
+// chunk, into an id.
+template <typename IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto ChunkIndicesToId(int chunk, int pos) -> IdT {
+  constexpr auto LowBits = PlatformChunkCapacityBits<IdT>();
+  // We can only use 31 bits in total because the sign bit is used for making
+  // negative ids which are not found in the ValueStore.
+  constexpr auto HighBits = 31 - LowBits;
+  // This routine is especially hot and the check here relatively expensive for
+  // the value provided, so only do this in debug builds to make tracking down
+  // issues easier.
+  CARBON_DCHECK(chunk < (1 << HighBits), "Id overflow (high bits)");
+  CARBON_DCHECK(pos < (1 << LowBits), "Id overflow (low bits)");
+  return IdT((chunk << LowBits) + pos);
+}
+
+// A chunk of `ValueType`s which has a fixed capacity, but variable size, and is
+// an iterable range.
+template <typename IdT, class ValueType>
+class ValueStoreChunk {
+ public:
+  static constexpr auto Capacity = Internal::PlatformChunkCapacitySize<IdT>();
+  static constexpr auto CapacityBytes =
+      Capacity * sizeof(typename IdT::ValueType);
+
+  ValueStoreChunk()
```

Comment:
Style: we've been doing explicit constructors by default

```suggestion
  explicit ValueStoreChunk()
```

Reply:
Done